packages/notebook/src/browser/view/notebook-main-toolbar.tsx

```diff
@@ -97,18 +114,64 @@ export class NotebookMainToolbar extends React.Component<NotebookMainToolbarProp
+            this.setState({ ...this.state, numberOfHiddenItems: this.state.numberOfHiddenItems + 1 });
+            this.lastGapElementWidth = this.gapElement.getBoundingClientRect().width;
+        } else if (this.gapElement && this.gapElement.getBoundingClientRect().width > this.lastGapElementWidth) {
+            this.setState({ ...this.state, numberOfHiddenItems: this.state.numberOfHiddenItems - 1 });
+            this.lastGapElementWidth = this.gapElement.getBoundingClientRect().width;
+        }
+    }
+
+    protected renderContextMenu(event: MouseEvent, menuItems: readonly MenuNode[]): void {
+        const hiddenItems = menuItems.slice(menuItems.length - this.state.numberOfHiddenItems);
+        const contextMenu = this.props.menuRegistry.getMenu([NotebookMenus.NOTEBOOK_MAIN_TOOLBAR_HIDDEN_ITEMS_CONTEXT_MENU]);
+
+        contextMenu.children.map(item => item.id).forEach(id => contextMenu.removeNode(id));
+        hiddenItems.forEach(item => contextMenu.addNode(item));
+
+        this.props.contextMenuRenderer.render({
+            anchor: event,
+            menuPath: [NotebookMenus.NOTEBOOK_MAIN_TOOLBAR_HIDDEN_ITEMS_CONTEXT_MENU],
+            context: this.props.editorNode,
+            args: [this.props.notebookModel.uri]
+        });
+    }
+
     override render(): React.ReactNode {
+        const menuItems = this.getMenuItems();
         return <div className='theia-notebook-main-toolbar'>
-            {this.getMenuItems().map(item => this.renderMenuItem(item))}
-            <div style={{ flexGrow: 1 }}></div>
+            {menuItems.slice(0, menuItems.length - this.state.numberOfHiddenItems).map(item => this.renderMenuItem(item))}
+            {
+                this.state.numberOfHiddenItems > 0 &&
+                <span className={`${codicon('ellipsis')} action-label theia-notebook-main-toolbar-item`} onClick={e => this.renderContextMenu(e.nativeEvent, menuItems)} />
+            }
+            <div ref={element => this.gapElementChanged(element)} style={{ flexGrow: 1 }}></div>
             <div className='theia-notebook-main-toolbar-item action-label'
                 onClick={() => this.props.commandRegistry.executeCommand(NotebookCommands.SELECT_KERNEL_COMMAND.id, this.props.notebookModel)}>
                 <span className={codicon('server-environment')} />
                 <span className=' theia-notebook-main-toolbar-item-text'>
                     {this.state.selectedKernelLabel ?? nls.localizeByDefault('Select Kernel')}
                 </span>
             </div>
-        </div>;
+        </div >;
+    }
+
+    protected gapElementChanged(element: HTMLDivElement | null): void {
+        this.gapElement = element ?? undefined;
+        if (this.gapElement) {
+            this.lastGapElementWidth = this.gapElement.getBoundingClientRect().width;
+            new ResizeObserver(() => this.calculateItemsToHide()).observe(this.gapElement);
```

Comment:
Suggestion: This code leaks the `ResizeObserver`. Every time the gap element is changed, another resize observer is added, which leads to a lot of calls to `calculateItemsToHide`. After changing the size of the window a few times, I already get more than 50+ calls to `calculateItemsToHide` by just maximizing the window.

The component should have a single resize observer that uses `observe` and `unobserve` as expected by the API.

Reply:
true should be fixed now